sslr-core/src/main/java/com/sonar/sslr/api/typed/GrammarBuilder.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * SonarSource Language Recognizer
+ * Copyright (C) 2010 SonarSource
+ * dev@sonar.codehaus.org
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package com.sonar.sslr.api.typed;
+
+import com.google.common.annotations.Beta;
+import org.sonar.sslr.grammar.GrammarRuleKey;
+
+import java.util.List;
+
+@Beta
+public interface GrammarBuilder<N, T> {
+
+  <U> NonterminalBuilder<U> nonterminal();
+
+  <U> NonterminalBuilder<U> nonterminal(GrammarRuleKey ruleKey);
+
+  <U> U firstOf(U... methods);
+
+  <U> Optional<U> optional(U method);
+
+  <U> List<U> oneOrMore(U method);
+
+  <U> Optional<List<U>> zeroOrMore(U method);
+
+  N invokeRule(GrammarRuleKey ruleKey);
```

Comment:
not convinced - this method was designed to return AstNode, and is not expected to be used in grammars that have been fully migrated (the master of Java should have fixed this) - but let's discuss